src/modelframe.jl

```diff
@@ -1,261 +1,102 @@
-```
-
-"""
-mutable struct ModelFrame
-    df::AbstractDataFrame
-    terms::Terms
-    nonmissing::BitArray
-    ## mapping from df keys to contrasts matrices
-    contrasts::Dict{Symbol, ContrastsMatrix}
+mutable struct ModelFrame{D,M}
+    f::FormulaTerm
+    schema
+    data::D
+    model::Type{M}
 end
 
-is_categorical(::AbstractArray{<:Union{Missing, Real}}) = false
-is_categorical(::AbstractArray) = true
-
-## Check for non-redundancy of columns.  For instance, if x is a factor with two
-## levels, it should be expanded into two columns in y~0+x but only one column
-## in y~1+x.  The default is the rank-reduced form (contrasts for n levels only
-## produce n-1 columns).  In general, an evaluation term x within a term
-## x&y&... needs to be "promoted" to full rank if y&... hasn't already been
-## included (either explicitly in the Terms or implicitly by promoting another
-## term like z in z&y&...).
-##
-## This modifies the Terms, setting `trms.is_non_redundant = true` for all non-
-## redundant evaluation terms.
-function check_non_redundancy!(trms::Terms, df::AbstractDataFrame)
-
-    (n_eterms, n_terms) = size(trms.factors)
 
-    encountered_columns = Vector{eltype(trms.factors)}[]
 
-    if trms.intercept
-        push!(encountered_columns, zeros(eltype(trms.factors), n_eterms))
+## copied from DataFrames:
+function _nonmissing!(res, col)
+    # workaround until JuliaLang/julia#21256 is fixed
+    eltype(col) >: Missing || return
+    
+    @inbounds for (i, el) in enumerate(col)
+        res[i] &= !ismissing(el)
     end
+end
 
-    for i_term in 1:n_terms
-        for i_eterm in 1:n_eterms
-            ## only need to check eterms that are included and can be promoted
-            ## (e.g., categorical variables that expand to multiple mm columns)
-            if Bool(trms.factors[i_eterm, i_term]) && is_categorical(df[trms.eterms[i_eterm]])
-                dropped = trms.factors[:,i_term]
-                dropped[i_eterm] = 0
-
-                if dropped ∉ encountered_columns
-                    trms.is_non_redundant[i_eterm, i_term] = true
-                    push!(encountered_columns, dropped)
-                end
-
-            end
-        end
-        ## once we've checked all the eterms in this term, add it to the list
-        ## of encountered terms/columns
-        push!(encountered_columns, view(trms.factors, :, i_term))
+function _nonmissing!(res, col::CategoricalArray{>: Missing})
+    for (i, el) in enumerate(col.refs)
+        res[i] &= el > 0
     end
-
-    return trms.is_non_redundant
 end
 
-const DEFAULT_CONTRASTS = DummyCoding
-
-# get unique values, honoring levels order
-_unique(x::AbstractCategoricalArray) = intersect(levels(x), unique(x))
-_unique(x::AbstractCategoricalArray{T}) where {T>:Missing} =
-    convert(Array{Missings.T(T)}, filter!(!ismissing, intersect(levels(x), unique(x))))
+_select(d::ColumnTable, cols::NTuple{N,Symbol} where N) = NamedTuple{cols}(d)
```

Comment:
This method seems to be called only by the helper below. Probably better remove one of them and handle the conversion from vector to tuple directly in the caller.